// ovsx/lib/util.js
"use strict";
/********************************************************************************
 * Copyright (c) 2019 TypeFox and others
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License v. 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0
 ********************************************************************************/
Object.defineProperty(exports, "__esModule", { value: true });
exports.promisify = void 0;
exports.addEnvOptions = addEnvOptions;
exports.matchExtensionId = matchExtensionId;
exports.optionalStat = optionalStat;
exports.makeDirs = makeDirs;
exports.createTempFile = createTempFile;
exports.rejectError = rejectError;
exports.handleError = handleError;
exports.statusError = statusError;
exports.readFile = readFile;
exports.readManifest = readManifest;
exports.validateManifest = validateManifest;
exports.writeFile = writeFile;
exports.writeManifest = writeManifest;
exports.getUserInput = getUserInput;
exports.getUserChoice = getUserChoice;
const fs = require("fs");
const path = require("path");
const tmp = require("tmp");
const readline = require("readline");
var util_1 = require("util");
Object.defineProperty(exports, "promisify", { enumerable: true, get: function () { return util_1.promisify; } });
function addEnvOptions(options) {
    options.registryUrl ?? (options.registryUrl = process.env.OVSX_REGISTRY_URL);
    options.pat ?? (options.pat = process.env.OVSX_PAT);
    options.username ?? (options.username = process.env.OVSX_USERNAME);
    options.password ?? (options.password = process.env.OVSX_PASSWORD);
}
function matchExtensionId(id) {
    return /^([\w-]+)(?:\.|\/)([\w-]+)$/.exec(id);
}
function optionalStat(path) {
    return new Promise((resolve, reject) => {
        fs.stat(path, (err, stats) => resolve(stats));
    });
}
function makeDirs(path) {
    return new Promise((resolve, reject) => {
        if (fs.existsSync(path)) {
            resolve();
        }
        else {
            fs.mkdir(path, { recursive: true }, (err) => {
                if (err)
                    reject(err);
                else
                    resolve();
            });
        }
    });
}
function createTempFile(options) {
    return new Promise((resolve, reject) => {
        tmp.tmpName(options, (err, name) => {
            if (err)
                reject(err);
            else
                resolve(name);
        });
    });
}
function rejectError(err) {
    const reason = err instanceof Error ? err : new Error(String(err));
    return Promise.reject(reason);
}
function handleError(debug, additionalMessage, exit = true) {
    return reason => {
        if (reason instanceof Error && !debug) {
            console.error(`\u274c  ${reason.message}`);
            if (additionalMessage) {
                console.error(additionalMessage);
            }
        }
        else if (typeof reason === 'string') {
            console.error(`\u274c  ${reason}`);
        }
        else if (reason !== undefined) {
            console.error(reason);
        }
        else {
            console.error('An unknown error occurred.');
        }
        if (exit) {
            process.exit(1);
        }
    };
}
function statusError(response) {
    if (response.statusMessage)
        return new Error(`The server responded with status ${response.statusCode}: ${response.statusMessage}`);
    else
        return new Error(`The server responded with status ${response.statusCode}.`);
}
function readFile(name, packagePath, encoding = 'utf-8') {
    return new Promise((resolve, reject) => {
        fs.readFile(path.join(packagePath ?? process.cwd(), name), { encoding }, (err, content) => {
            if (err) {
                reject(err);
            }
            else {
                resolve(content);
            }
        });
    });
}
async function readManifest(packagePath) {
    const content = await readFile('package.json', packagePath);
    return JSON.parse(content);
}
function validateManifest(manifest) {
    if (!manifest.publisher) {
        throw new Error("Missing required field 'publisher'.");
    }
    if (!manifest.name) {
        throw new Error("Missing required field 'name'.");
    }
    if (!manifest.version) {
        throw new Error("Missing required field 'version'.");
    }
}
function writeFile(name, content, packagePath, encoding = 'utf-8') {
    return new Promise((resolve, reject) => {
        fs.writeFile(path.join(packagePath ?? process.cwd(), name), content, { encoding }, (err) => {
            if (err) {
                reject(err);
            }
            else {
                resolve();
            }
        });
    });
}
function writeManifest(manifest, packagePath) {
    const content = JSON.stringify(manifest, null, 4);
    return writeFile('package.json', content, packagePath);
}
function getUserInput(text) {
    return new Promise(resolve => {
        const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
        rl.question(text, answer => {
            resolve(answer);
            rl.close();
        });
    });
}
async function getUserChoice(text, values, defaultValue, lowerCase = true) {
    const prompt = text + '\n' + values.map(v => v === defaultValue ? `[${v}]` : v).join('/') + ': ';
    const answer = await getUserInput(prompt);
    if (!answer) {
        return defaultValue;
    }
    const lcAnswer = lowerCase ? answer.toLowerCase() : answer;
    for (const value of values) {
        if (value.startsWith(lcAnswer)) {
            return value;
        }
    }
    return defaultValue;
}
